cpp/src/umap/fuzzy_simpl_set/naive.cuh

```diff
@@ -372,6 +339,43 @@ void launcher(nnz_t n,
       return res;
     },
     stream);
+}
+
+/**
+ * Given a set of X, a neighborhood size, and a measure of distance, compute
+ * the fuzzy simplicial set (here represented as a fuzzy graph in the form of
+ * a sparse coo matrix) associated to the data. This is done by locally
+ * approximating geodesic (manifold surface) distance at each point, creating
+ * a fuzzy simplicial set for each such point, and then combining all the local
+ * fuzzy simplicial sets into a global one via a fuzzy union.
+ *
+ * @param n the number of rows/elements in X
+ * @param knn_indices indexes of knn search
+ * @param knn_dists distances of knn search
+ * @param n_neighbors number of neighbors in knn search arrays
+ * @param out The output COO sparse matrix
+ * @param params UMAPParams config object
+ * @param stream cuda stream to use for device operations
+ */
+template <typename value_t, typename value_idx, typename nnz_t, int TPB_X>
+void launcher(nnz_t n,
+              const value_idx* knn_indices,
+              const value_t* knn_dists,
+              int n_neighbors,
+              raft::sparse::COO<value_t>* out,
+              UMAPParams* params,
+              cudaStream_t stream)
+{
+  /* Nested scope so `strengths` is dropped before the `coo_sort` call to
+   * reduce device memory usage. */
+  {
+    raft::sparse::COO<value_t> strengths(stream, n * n_neighbors, n, n);
+
+    compute_membership_strength<value_t, value_idx, nnz_t, TPB_X>(
+      n, knn_indices, knn_dists, n_neighbors, &strengths, params, stream);
+
+    symmetrize<value_t>(&strengths, out, params->set_op_mix_ratio, stream);
+  }
```

Comment:
Could you add an in-line comment for the closing brackets of scopes used to define lifetime? Something like this?
```suggestion
  }  // End strengths scope
```